microovn/ovn/commands.go

```diff
@@ -136,14 +139,46 @@ func VSCtl(s *state.State, args ...string) (string, error) {
 		return "", err
 	}
 
-	err = waitForDBConnected(s, dbSpec, defaultDBConnectWait)
+	err = waitForDBState(s, dbSpec, OvsdbConnected, defaultDBConnectWait)
 	if err != nil {
 		return "", err
 	}
 
 	return shared.RunCommandContext(s.Context, "ovs-vsctl", args...)
 }
 
+// AppCtl is a convenience function that wraps execution of 'ovn-appctl' command. It requires argument
+// 'target' which will be substituted to the '-t' argument of 'ovn-appctl'. Rest of the 'args' will be passed
+// to the ovn-appctl unchanged.
+func AppCtl(s *state.State, target string, args ...string) (string, error) {
+	arguments := []string{"-t", target}
+	arguments = append(arguments, args...)
+	return shared.RunCommandContext(
+		s.Context,
+		"ovn-appctl",
+		arguments...,
+	)
+}
+
```

Comment:
`ovn-controller` uses a a control socket with random number in the filename (i.e. `ovn-controller.3376923.ctl=`) unlike OVN SB or NB processes which use predictable names like `ovnnb_db.ctl=`. However `ovs-appctl` can figure out correct control socket used by `ovn-controller` if env variable `OVS_RUNDIR` is properly set. :point_down:

Reply:
The OVN project provides its own appctl variant, `ovn-appctl`, which will have the appropriate defaults to handle this communication without the extra arguments.

So can't you just use the above `AppCtl` function for this, and if not, there may be environment variables not set up properly?

Example from a classic deployment:
```
$ sudo ovn-appctl list-commands
The available commands are:
  connection-status       
  coverage/read-counter   COUNTER
  coverage/show           
  ct-zone-list            
...
$ sudo ovn-appctl -t ovn-controller list-commands
The available commands are:
  connection-status       
  coverage/read-counter   COUNTER
  coverage/show           
  ct-zone-list            
```

Reply:
`ovn-appctl` and `ovs-appctl` both behave exactly the same in regards to the "ovn-controller". You can specify `-t controller` for both of them and the auto-discovery mechanism will work as long as runtime directory is properly set.
* `$OVS_RUNDIR` for `ovs-appctl`
* `$OVN_RUNDIR` for `ovn-appctl`

I chose to use `ovs-appctl` just because that's what's mentioned in [the docs](https://man7.org/linux/man-pages/man8/ovn-controller.8.html) but I also tested it and both commands work the same. At the end of the day, we need to set one of these two env vars and invoke matching command with `-t controller`.

I also though about adding new wrapper that would set "rundir" variable to chassis' runtime directory but that seemed like a bigger disruption to the UX that would have to be properly documented and maintained.

Reply:
> `ovn-appctl` and `ovs-appctl` both behave exactly the same in regards to the "ovn-controller". You can specify `-t controller` for both of them and the auto-discovery mechanism will work as long as runtime directory is properly set.
> 
>     * `$OVS_RUNDIR` for `ovs-appctl`
> 
>     * `$OVN_RUNDIR` for `ovn-appctl`
> 
> 
> I chose to use `ovs-appctl` just because that's what's mentioned in [the docs](https://man7.org/linux/man-pages/man8/ovn-controller.8.html) but I also tested it and both commands work the same. At the end of the day, we need to set one of these two env vars and invoke matching command with `-t controller`.

The `ovs-appctl` in the docs is much likely a left-over from OVS/OVN splitting into separate projects.

When not having these tools in a snap, the build time configuration would make the tool dtrt. on its own, so the environment variable is a snap problem.

> I also though about adding new wrapper that would set "rundir" variable to chassis' runtime directory but that seemed like a bigger disruption to the UX that would have to be properly documented and maintained.

Which UX are you thinking of here developer UX, the `Appctl` function is not exposed to the CLI right?

I think it would be more appropriate to have `OvsAppCtl` and `OvnAppCtl` functions that set the environment variable appropriately. These tools have a wide range of uses and we may need generic access to them later, rater than having a super specialized `ControllerCtl`.

Reply:
>> I also though about adding new wrapper that would set "rundir" variable to chassis' runtime directory but that seemed like a bigger disruption to the UX that would have to be properly documented and maintained.

> Which UX are you thinking of here, developer UX? The Appctl function is not exposed to the CLI right?

I thought about creating the snap/cli wrapper, like `microovn.controller-ctl`, but I decided against it.

> I think it would be more appropriate to have OvsAppCtl and OvnAppCtl functions that set the environment variable appropriately. These tools have a wide range of uses and we may need generic access to them later, rather than having a super specialized ControllerCtl.

I agree with you that we should have general `OvsAppCtl` and `OvnAppCtl` However if we were to set runtime directories for these commands automatically we'd set
* `$SNAP_COMMON/run/central` for `OvnAppCtl`
* `$SNAP_COMMON/run/switch` for `OvsAppCtl`
But neither of these is appropriate when we want to target controller which needs `$SNAP_COMMON/run/chassis`. That's why I thought that adding the third function would be helpful. We can also make the wrappers really bare-bone and force user to pass all required arguments but I think that detracts a bit from usefulness.

Reply:
Sorry. I kinda forgot to conclude my thought in previous comment. My proposal is that we have three functions
* Generic `OvnAppCtl`
* Generic `OvsAppCtl` (there wasn't a use case for this one yet so it doesn't exist)
* "Helper" `ControllerCtl` that uses either `ovs-appctl` or `ovn-appctl` and sets proper runtime directory for targeting controller.

What do you think about this approach?

Reply:
> Sorry. I kinda forgot to conclude my thought in previous comment. My proposal is that we have three functions
> 
>     * Generic `OvnAppCtl`
> 
>     * Generic `OvsAppCtl` (there wasn't a use case for this one yet so it doesn't exist)

:+1: 
 
>     * "Helper" `ControllerCtl` that uses either `ovs-appctl` or `ovn-appctl` and sets proper runtime directory for targeting controller.

I think we should avoid this, and instead figure out why this does not work out of the box.

Having thought some more about your comment about UX, the end user of `microovn.ovs-appctl` and `microovn.ovn-appctl` would expect those tools to make the appropriate decisions by default. Why are they currently not doing so?

Looking at microovn/snapcraft/ovn.env sourced by the snap `ovn-appctl` script:
https://github.com/canonical/microovn/blob/407a98a32ff5e8c8e8e1e0ad5d933bb5c159f44f/snapcraft/ovn.env#L9

Perhaps that is causing the issue?

The ovs-* commands already set `OVS_RUNDIR` here:
https://github.com/canonical/microovn/blob/407a98a32ff5e8c8e8e1e0ad5d933bb5c159f44f/snapcraft/commands/ovs#L2

Reply:
>  export OVN_RUNDIR="${SNAP_COMMON}/run/switch/"

This is indeed strange and looks like a typo?

> I think we should avoid this, and instead figure out why this does not work out of the box.

We have two CLI tools

* `ovn-appctl`
* `ovs-appctl`

And three runtime directories

`$SNAP_COMMON/run/central`
`$SNAP_COMMON/run/switch`
`$SNAP_COMMON/run/chassis`

So I think we have two options:

* Add 3rd wrapper (be it in code, or as a CLI wrapper/script) and each can default to correct runtime dir
* Don't add 3rd wrapper but make "runtime dir" a required argument for both `OvsAppCtl` and `OvnAppCtl`.

The magic `-t controller` works by looking for `ovn-controller.pid` file in `$OVS_RUNDIR` or `$OVN_RUNDIR` (based on whether you are running `ovs-appctl` or `ovn-appctl` respectively). It then takes the pid and looks up socket associated with that process.
So another alternative is to just cancel `$SNAP_COMMON/run/chassis/` and have controller pid and socket files in `$SNAP_COMMON/run/switch`. This way the `ovs-appctl` would work out of the box with `-t controller`. However that looked like a pretty big structure change to me so I didn't go for it.

Reply:
Ooh, it did not occur to me that we had split the runtime directories across three paths, I see the predicament now.

OK, you're suggestion is then fine, lets go with it.

In a future change, I think we should merge `$SNAP_COMMON/run/central` and `$SNAP_COMMON/run/chassis` into `$SNAP_COMMON/run/ovn`, unless there are strong counter arguments against it. That would rid us of this problem and also retain compatibility with upstream intended use of these tools.

What do you think?

Reply:
Yup, I fully agree. If there are no blockers for merging of runtime dirs, we should do it.